### ISSUE TITLE
[no-unused-vars] vars is used but why warning still? 

### ISSUE THREAD

**Issue by MariuzM** [typescript-eslint/typescript-eslint#1675]
I'm getting `App` and `React` `eslint(@typescript-eslint/no-unused-vars)` warnings but it's being used in render method. Same issue happens in my other components but for simplicity sake I’ve just used this one. 

index.tsx
```js
import * as React from 'react'
import * as ReactDOM from 'react-dom'

import App from './App'

ReactDOM.render(<App />, document.getElementById('root'))
```
.eslintrc.js
```js
module.exports = {
  parser: '@typescript-eslint/parser',
  extends: [
    'plugin:@typescript-eslint/recommended',
    'prettier/@typescript-eslint',
    'plugin:prettier/recommended', // Enables eslint-plugin-prettier and displays prettier errors as ESLint errors. Make sure this is always the last configuration in the extends array.
  ],
  parserOptions: {
    sourceType: 'module', // Allows for the use of imports
    ecmaVersion: 2018, // Allows for the parsing of modern ECMAScript features
    // ecmaFeatures: { jsx: true },
  },
  // overrides: [
  //   {
  //     files: ['*.ts', '*.tsx'],
  //     rules: {
  //       '@typescript-eslint/no-unused-vars': false,
  //     },
  //   },
  // ],
  // plugins: ['typescript', 'react', 'standard', 'prettier'],
}
```

.babelrc
```json
{
  "presets": ["@babel/env", "@babel/typescript"],
  "plugins": ["@babel/proposal-class-properties"]
}
```

tsconfig.json
```json
{
  "compilerOptions": {
    // "target": "esnext",
    "target": "es6",
    "module": "commonjs",
    "jsx": "react",
    "watch": true,
    "lib": ["es6", "dom"],
    "outDir": "./dist",
    "sourceMap": true,
    "noImplicitAny": true,
    "moduleResolution": "node",
    "strict": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "esModuleInterop": true
  }
}
```
package.json / dev
```json
	"devDependencies": {
		"@babel/core": "^7.8.6",
		"@babel/plugin-proposal-class-properties": "^7.8.3",
		"@babel/preset-react": "^7.8.3",
		"@babel/preset-typescript": "^7.8.3",
		"@types/react": "^16.9.23",
		"@types/react-dom": "^16.9.5",
		"@typescript-eslint/eslint-plugin": "^2.22.0",
		"@typescript-eslint/parser": "^2.22.0",
		"babel-loader": "^8.0.6",
		"eslint": "^6.8.0",
		"eslint-config-prettier": "^6.10.0",
		"eslint-plugin-import": "^2.20.1",
		"eslint-plugin-jsx-a11y": "^6.2.3",
		"eslint-plugin-prettier": "^3.1.2",
		"eslint-plugin-react": "^7.18.3",
		"eslint-plugin-react-hooks": "^1.7.0",
		"parcel-bundler": "^1.12.4",
		"prettier": "^1.19.1",
		"source-map-loader": "^0.2.4",
		"typescript": "^3.8.3"
	},
```

**Comment by bradzacher** [comment]
The rule does not understand jsx usage semantics.

If you want the rule to understand that jsx uses a variable, you must use the rules from the `eslint-plugin-react` package.

**Comment by plusgut** [comment]
@bradzacher How is that working as intended? Typescript transforms that jsx to something where the variable *is* used.
That has nothing to do with react, but with how typescript handles jsx.

**Comment by bradzacher** [comment]
The base rule is built by ESLint, and they do not want to support JSX syntax in the rule for various reasons which are completely understandable.

This is why `eslint-plugin-react` created their rules that teach the rule how to understand JSX semantics.

Since then, JSX has grown to be used outside of react, but that rule in that plugin still works just the same, even if you don't use react.

Our version of the rule simply teaches the rule about typescript syntax, and builds on top of the base rule. We are not going to duplicate the work done by that plugin, because there is no need.

If you want a version that handles JSX, then try our `no-unused-vars-experimental` rule, which has an entirely different implementation based off of internal TS code.
